core/util.py

```diff
@@ -89,3 +93,33 @@ def group_by_first_element(tuples: list[tuple]) -> dict[str, list[tuple | Any]]:
         key: [t[1:] if len(t[1:]) > 1 else t[1] for t in group] for key, group in groups
     }  # Map groups into a dictionary, removing the first index from the group as that's the key
     return nested
+
+
+def get_sheet_section_and_column(validation_failure: Any) -> tuple[str, str, str]:
+    """Returns the sheet and section for a given validation failure."""
+    sheet = INTERNAL_TABLE_TO_FORM_TAB[validation_failure.sheet]
+    column, section = INTERNAL_COLUMN_TO_FORM_COLUMN_AND_SECTION[validation_failure.column]
+
+    # additional logic for outcomes to differentiate between footfall and non-footfall
+    if sheet == "Outcomes" and hasattr(validation_failure, "row_values"):
+        if validation_failure.row_values[4] == "Year-on-year % change in monthly footfall":
+            section = "Footfall Indicator"
+
+    # additional logic for risk location
+    if sheet == "Risk Register" and hasattr(validation_failure, "row_values"):
+        if validation_failure.row_values[1]:
+            # project risk
+            section = f"Project {int(validation_failure.row_values[1].split('-')[2])} Risks"
+        else:
+            # programme risk
+            section = "Programme Risks"
+
+    # additional logic if column is 'Question'
+    if validation_failure.column == "Question" and sheet == "Project Admin":
+        section = "Place Details"
+
+    # additional logic if Project Outputs as always has same section
+    if sheet == "Project Outputs":
+        section = "Project Outputs"
+
+    return sheet, section, column
```

Comment:
The type hint for the validation_failure param in this function should be `TFUCFailureMessage` rather than `Any`.

I think this raises an issue now though. If we want a generic function that will work with different types of validation failures then we should standardise the `TFUCFailureMessage` objects to have a standard interface.

If we elevate the `column`, `sheet` and `row_values` out into the parent class `TFUCFailureMessage` we can be sure that a function like this can work with the objects it's passed and won't need to make checks like above where we do `hasattr` because we'll know the interface for all objects passed to the function.

This raises another issue though - e.g. `NonNullableConstraintFailure` cannot have `row_values` because the precision of the error is only at the column level, rather than cell. Does this mean we would change the validation capture logic to be at the cel level?

Another example is that `NonUniqueCompositeKeyFailure` uses `cols` rather than a single column attribute. 

I think this begs the question - do we standardise these 4 failures so a generic function like this can be robust - or do we just have individual implementations of this logic for each error? I'm not sure the logic within this is actually as generic as it seems - i.e. it covers so many individual edge cases and not all errors can occur on all sheets. Building for reusability and genericism is good practice as a general rule but I think we need to consider this? What do you think?

Reply:
The lack of ability to extract row_values from a NonUniqueCompositeKeyFailure does present some issues in combining all of the above logic into one piece of reusable code. It seemed like a sensible idea, but given how hyper specific each of these errors are going to be (relevant to different columns and sheets, returning different pieces of information from which we reverse engineer back to the original spreadsheet) I think I'll revert to getting the column, sheet, and section bespoke for each function.

Reply:
And then all of the functions are complete we can assess how much is reusable and potentially recombine those chunks of code into a function like the above.

Reply:
sounds good to me